pkg/generate/code/resource_reference.go

```diff
@@ -0,0 +1,87 @@
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//     http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package code
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/aws-controllers-k8s/code-generator/pkg/model"
+)
+
+// ReferenceFieldsValidation produces the go code to validate reference field and
+// corresponding identifier field.
+// Sample code:
+// if ko.Spec.APIIDRef != nil && ko.Spec.APIID != nil {
+//		return ackerr.ResourceReferenceAndIDNotSupportedFor("APIID", "APIIDRef")
+//	}
+//	if ko.Spec.APIIDRef == nil && ko.Spec.APIID == nil {
+//		return ackerr.ResourceReferenceOrIDRequiredFor("APIID", "APIIDRef")
+//	}
+func ReferenceFieldsValidation(
+	crd *model.CRD,
+	sourceVarName string,
+	referenceFieldSuffix string,
+	indentLevel int,
+) string {
+	out := ""
+	for _, field := range crd.Fields {
+		if field.HasReference() {
+			indent := strings.Repeat("\t", indentLevel)
+			// Validation to make sure both target field and reference are
+			// not present at the same time in desired resource
+			out += fmt.Sprintf("%sif %s.Spec.%s%s != nil"+
+				" && %s.Spec.%s != nil {\n", indent, sourceVarName, field.Names.Camel,
+				referenceFieldSuffix, sourceVarName, field.Names.Camel)
+			out += fmt.Sprintf("%s\treturn "+
+				"ackerr.ResourceReferenceAndIDNotSupportedFor(\"%s\", \"%s%s\")\n",
+				indent,field.Names.Camel, field.Names.Camel, referenceFieldSuffix)
+			out += fmt.Sprintf("%s}\n", indent)
+
+			// If the field is required, make sure either Ref or original
+			// field is present in the resource
+			if field.IsRequired() {
+				out += fmt.Sprintf("%sif %s.Spec.%s%s == nil &&"+
+					" %s.Spec.%s == nil {\n", indent, sourceVarName,
+					field.Names.Camel, referenceFieldSuffix, sourceVarName,
+					field.Names.Camel)
+				out += fmt.Sprintf("%s\treturn "+
+					"ackerr.ResourceReferenceOrIDRequiredFor(\"%s\", \"%s%s\")\n",
+					indent, field.Names.Camel, field.Names.Camel,
+					referenceFieldSuffix)
+				out += fmt.Sprintf("%s}\n", indent)
+			}
+		}
+	}
+	return out
+}
```

Comment:
I am OK with the above but still think it's a bit too ACK-specific (with the ackerr error messages). I'd like eventually for these Go code generating functions to be smaller and more generic but won't hold this particular PR up for that long-term goal.